Lab03 - Letter frequency

n0npax opened the pull request:
Fixes #207 

Review of colleague's PR #200

#### Changes proposed in this PR:

Implement solution for lab03

* letterFrequency
* tests

Commit:
Complete lab03 - Letter frequency

Map letters to frequency
Return ordered list of latters

rokane commented on 03_letters/n0npax/main.go:
```diff
@@ -0,0 +1,52 @@
+	for i := range ss {
+		freqMap[ss[i]]++
+	}
+	return
```
Interesting, I haven't seen this too often. I would normally go down the path of explicitly returning the `freqMap` variable as I feel it helps with readability, and is clear what you are returning. But I don't hate, it is growing on me.

juliaogris replied:
Avoid named result parameters and naked returns unless you have got good reason  (source [golang wiki](https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters)) - which I believe you don't here.

n0npax replied:
fixed. Will avoid this in the future. Thanks!

rokane commented on 03_letters/n0npax/main.go:
```diff
@@ -0,0 +1,52 @@
+	freqMap = map[rune]int{}
+	ss := []rune(s)
+	for i := range ss {
+		freqMap[ss[i]]++
```
Nice work 👍

rokane commented on 03_letters/n0npax/main.go:
```diff
@@ -0,0 +1,52 @@
+
+type letterFreq struct {
+	letter    rune
+	occurency int
```
Can you please rename this to `freq` or `count`

```go
type letterFreq struct {
  letter  rune
  count int
}
```

n0npax replied:
done. Thanks for that

rokane commented on 03_letters/n0npax/main_test.go:
```diff
@@ -0,0 +1,109 @@
+		for k, v := range result {
+			if v != expected[k] {
+				for k := range result {
+					if expected[k] != result[k] {
```
I think that this is starting to get quite confusing and difficult to read. I don't think it is necessary to do the final iteration over the result. Can you please remove the double nested loop like below?

```go
for k, count := range result {
  if _, ok := expected[k]; !ok {
    t.Errorf("Found unexpected key: %c", k)
  }
  if count != expected[k] {
    t.Errorf("Incorrect count for key %c. Found: %d, Expected %d", k, count, expected[k])
  }
}
```

n0npax replied:
Thanks for that. Fixed :)

rokane reviewed (changes requested):
Looking really good. I like the way you are breaking things down and abstracting your test suites out. If you can please update according to the two comments I have provided, then this will be right for merging 👍

Commit:
Fix miscellaneous according to CR

Fix overcomplicated logic in tests
Rename variable
Split tests into positive and negative

juliaogris commented on 03_letters/n0npax/main.go:
```diff
@@ -0,0 +1,52 @@
+}
+
+func sortLetters(m map[rune]int) (sortedLetters []string) {
+
```
unnecessary blank line

n0npax replied:
fixed. Thanks

juliaogris commented on 03_letters/n0npax/main.go:
```diff
@@ -0,0 +1,52 @@
+	for i, lf := range freq {
+		sortedLetters[i] = lf.String()
+	}
+	return
```
avoid naked returns and named result parameters, as above.

n0npax replied:
Fixed. Thanks!

juliaogris commented on 03_letters/n0npax/main_test.go:
```diff
@@ -0,0 +1,109 @@
+	}{
+		"latinum":  {input: letterFreq{letter: 'a', count: 2}, expected: "a:2", positive: true},
+		"cyrillic": {input: letterFreq{letter: 'а', count: 2}, expected: "а:2", positive: true},
+		// cyrillic 'а' is different char than latin 'a'
```
neat!

juliaogris commented on 03_letters/n0npax/main_test.go:
```diff
@@ -0,0 +1,109 @@
+		"mixed": {input: letterFreq{letter: 'а', count: 2}, expected: "a:2", positive: false},
+	}
+	for name, test := range testCases {
+		t.Run(name, helperTestLetterFreqString(name, test.input, test.expected, test.positive))
```
Nit: this is a matter of style, but imo it is better to keep TestFuncs purely self contained module some test fixtures to generate input data. Helpers should not be used to create assertions.

I your case I'd just pull out a bunch of `NotEqual` and `Equal` test cases in two separate test functions and test tables I also recommend using `testify/assert` at this stage.

then your asserts could just by:
```go
for name, tc := range testCases {
     t.Run(name, func(t *testing.T) {
        got :=  tc.input.String()
        assert.Equal(t, tc.expected, got)
    } )
}
```
and in the not equal case:
```go
for name, tc := range testCases {
     t.Run(name, func(t *testing.T) {
        got :=  tc.input.String()
        assert.NotEqual(t, tc.notExpected, got)
    } )
}
```

n0npax replied:
fixed. Thanks for that!

juliaogris reviewed (dismissed):
A few more little nitpicks - overall looks really great!

Please fix up your last commit messages: `CR fixes` should be `Fix miscellaneous according to CR` or similar. Better yet: squash all commits into a single commit  with `git rebase -i ` or `git commit --amend` to update only the last commit and then `git push -f`.
If you do want separate commits, I recommend: `git commit --fixup <HASH>` where <HASH> is the commit SHA1 of the commit which is being fixed up; if you run `git rebase -i --autosquash master` it should do as you want.

rokane commented on 03_letters/n0npax/main_test.go:
```diff
@@ -0,0 +1,122 @@
+func helperTestLetter(input string, expected map[rune]int) func(*testing.T) {
+	return func(t *testing.T) {
+		result := letters(input)
+		for k, count := range result {
```
I just realised there is another case which you should guard against in this test, you should check the length of the result against the length of the expected. 
If there were a case where `result := { 'a': 1, 'b': 2 }` and `expected := { 'a': 1, 'b': 2, 'c': 3 }` the below wouldn't be able to catch it.

```go
result := letters(input)
if len(result) != len(expected) {
  t.Errorf("some error msg here ... ")
}
for k, count := range result {
  ...
}

n0npax replied:
Fixed. Thanks

rokane reviewed (changes requested):
Apologies, I should have picked up on that case the first time I reviewed this. Are you able to please include that one and then I will take another look first thing in the morning.

n0npax committed:
Merge branch 'master' into lab03
